pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"reflect"
+	"sync"
+
+	"time"
+
+	calicocache "github.com/projectcalico/k8s-policy/pkg/cache"
+	"github.com/projectcalico/k8s-policy/pkg/controllers/controller"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	corecache "k8s.io/client-go/tools/cache"
+
+	"fmt"
+)
+
+// This cache contains kubernetes node information keyed by etcd node-name.
+type cache struct {
+	sync.RWMutex
+	// The key for the cache's data is an etcd nodeName.
+	data map[string]string
+}
+
+// NodeController implements the Controller interface.  It is responsible for monitoring
+// kubernetes nodes and responding to delete events by removing them from the Calico datastore.
+// It keeps a cache of known calico nodes and their corresponding kubernetes nodes to
+// accomplish this.
+type NodeController struct {
+	indexer       corecache.Indexer
+	informer      corecache.Controller
+	nodeDataCache calicocache.ResourceCache
+	nodeCache *cache
+	calicoClient  *client.Client
+	k8sClientset  *kubernetes.Clientset
+}
+
+
+// NewNodeController Constructor for NodeController
+func NewNodeController(k8sClientset *kubernetes.Clientset, calicoClient *client.Client) controller.Controller {
+	cacheArgs := calicocache.ResourceCacheArgs{
```

Comment:
I think we want to pass some reconciler config here, similar to [what the pod controller does](https://github.com/projectcalico/kube-controllers/blob/master/pkg/controllers/pod/pod_controller.go#L80-L86)

Namely, we don't want the reconciler to send us events for when something exists in the Kubernetes API but not in the Calico datastore, because we know there's our workers can do about that and it's just going to generate extra work.

So, we at least want to set `DisableMissingInDatastore: true` for that case.
I think we can leave `DisableMissingInCache: false` because we do care about and can handle the case when something exists in the Calico datastore but not in the k8s API.
I think the default `DisableUpdateOnChange: false` is fine too because we'll never actually have values between cache / datastore differ, since it's always set to `""`.